internal/core/items/text.rs

```diff
@@ -571,18 +571,22 @@ impl TextInput {
             TextCursorDirection::NextLine => {
                 reset_preferred_x_pos = false;
 
-                let mut cursor_xy_pos =
-                    window.text_input_position_for_byte_offset(self, last_cursor_pos);
-                cursor_xy_pos.y += font_height * 1.5;
+                let cursor_rect =
+                    window.text_input_cursor_rect_for_byte_offset(self, last_cursor_pos);
+                let mut cursor_xy_pos = cursor_rect.center();
+
+                cursor_xy_pos.y += font_height;
                 cursor_xy_pos.x = self.preferred_x_pos.get();
                 window.text_input_byte_offset_for_position(self, cursor_xy_pos)
             }
             TextCursorDirection::PreviousLine => {
                 reset_preferred_x_pos = false;
 
-                let mut cursor_xy_pos =
-                    window.text_input_position_for_byte_offset(self, last_cursor_pos);
-                cursor_xy_pos.y -= font_height * 0.5;
+                let cursor_rect =
+                    window.text_input_cursor_rect_for_byte_offset(self, last_cursor_pos);
+                let mut cursor_xy_pos = cursor_rect.center();
+
+                cursor_xy_pos.y -= font_height;
```

Comment:
Here you also need to special case for the first line. If it becomes negative, we don't want to move the cursor.
Also on the last line we also don't want to move the cursor when going down